### ISSUE TITLE
Allow a limit of the number of active and persistent anchors

### ISSUE THREAD

**Issue by cabanier** [immersive-web/anchors#79]
Currently the spec does not mention any limit to the number of anchors a site can create.
On Quest, there is a significant additional load if a lot of anchors are active so we like to limit how many can be in a scene at a time.
For persistence, the browser can only create a limited set of persistent anchors across all origins. We'd like the spec to be updated so the browser is allowed to deleted older anchors to make space for new ones.

/agenda Allow a limit of the number of active and persistent anchors

**Comment by cabanier** [comment]
We discussed this in today's meeting.
The spec will be updated to enforce a guaranteed minimum number of anchors and allow the UA to limit them to a maximum.

**Comment by Jmnil-001** [comment]
When will persistent anchors be supported in Google Chrome?

**Comment by cabanier** [comment]
> When will persistent anchors be supported in Google Chrome?

I think we need to wait until Android supports them.

**Comment by Jmnil-001** [comment]
> > When will persistent anchors be supported in Google Chrome?
> 
> I think we need to wait until Android supports them.

Is there any Android device that supports persistent anchors now? If not, do you know approximately when Android devices will support this feature?

**Comment by cabanier** [comment]
Quest is Android based :-)
As for a general implementation, only Google knows...

**Comment by HyroVitalyProtago** [comment]
@cabanier how many anchors max on quest?

**Comment by cabanier** [comment]
> [@cabanier](https://github.com/cabanier) how many anchors max on quest?

a maximum of 8

**Comment by salmanmkc** [comment]
> > [@cabanier](https://github.com/cabanier) how many anchors max on quest?
> 
> a maximum of 8

I had ran into issues where I could only get about 6 on the meta quest 3, and I could try to clear it, however if that didn't work I would have to clear the browser memory in the settings

**Comment by cabanier** [comment]
> > > [@cabanier](https://github.com/cabanier) how many anchors max on quest?
> > 
> > 
> > a maximum of 8
> 
> I had ran into issues where I could only get about 6 on the meta quest 3, and I could try to clear it, however if that didn't work I would have to clear the browser memory in the settings

Isn't there a list of persistent anchors? You could delete them from there.

**Comment by salmanmkc** [comment]
> > > > [@cabanier](https://github.com/cabanier) how many anchors max on quest?
> > > 
> > > 
> > > a maximum of 8
> > 
> > 
> > I had ran into issues where I could only get about 6 on the meta quest 3, and I could try to clear it, however if that didn't work I would have to clear the browser memory in the settings
> 
> Isn't there a list of persistent anchors? You could delete them from there.

Yeah, that's what I was doing, reading `session.persistentAnchors` and calling `deletePersistentAnchor` on each one, behind a "Release" button in the demos.

It works while my own origin still names the handles. The trouble is the cases where it doesn't. A few things I logged on a Quest 3 while testing:

- `requestPersistentHandle` failed with "Maximum number of anchors reached!" at the same time as `persistentAnchors` returned an empty list, so there was nothing to delete and no way back from inside the page. That happened three separate times.
- The list had 9 entries at one point, and 3 of those were empty strings. `deletePersistentAnchor("")` rejects with `OperationError: Anchor deletion failed.`, so I filter blanks out now. An origin that had never saved an anchor still reported `[""]`.
- The list doesn't update within a session. I deleted 6 handles successfully and it still reported 9 straight after. The next session listed those same 6, and every one of them failed with `OperationError: Anchor deletion failed.`, presumably because they were already gone.
- Occasionally `deletePersistentAnchor` rejects with `InvalidStateError: Anchor is already being restored or deleted`, and then works on a retry.

Nothing I could do from the page got me out of the full state. It only came back after I cleared things on the device itself, and I didn't note down which setting did it.

The spec now says the UA may delete entries to make space ("if the maximum number of system anchors is reached for all maps for all origins, the user agent MAY free up an entry to make space for the newly requested anchor"). Is that implemented, or still on the list? Straight LRU would mean one origin can silently drop another's anchors, so a per origin quota might be closer to what's wanted, but either would beat being locked out by anchors you can't enumerate.

**Comment by cabanier** [comment]
> * `requestPersistentHandle` failed with "Maximum number of anchors reached!" at the same time as `persistentAnchors` returned an empty list, so there was nothing to delete and no way back from inside the page. That happened three separate times.

Interesting! Do you have a test case that reproduces this?

> * The list had 9 entries at one point, and 3 of those were empty strings. `deletePersistentAnchor("")` rejects with `OperationError: Anchor deletion failed.`, so I filter blanks out now. An origin that had never saved an anchor still reported `[""]`.

That was an error that was introduced a couple of months ago. Please ignore the empty strings for now. This issue will be fixed in an upcoming browser release.

> * The list doesn't update within a session. I deleted 6 handles successfully and it still reported 9 straight after. The next session listed those same 6, and every one of them failed with `OperationError: Anchor deletion failed.`, presumably because they were already gone.
> * Occasionally `deletePersistentAnchor` rejects with `InvalidStateError: Anchor is already being restored or deleted`, and then works on a retry.

hmm, I tried this but could not reproduce. Are you calling 'await' on each release?

> The spec now says the UA may delete entries to make space ("if the maximum number of system anchors is reached for all maps for all origins, the user agent MAY free up an entry to make space for the newly requested anchor"). Is that implemented, or still on the list? Straight LRU would mean one origin can silently drop another's anchors, so a per origin quota might be closer to what's wanted, but either would beat being locked out by anchors you can't enumerate.

This is not happening at the moment because I felt uncomfortable deleting a random anchor... Do you think I should?

**Comment by KutlayTunc** [comment]
Quest 3 (Horizon OS build UP1A.231005.007.A1 / 52433670036000520)
OculusBrowser/150.1.0.24.52.1046134268
Chrome/150.0.7871.224

Read live over adb + CDP 

Our page keeps one persistent anchor per session

1. Store full with nine listed entries, all blank: `requestPersistentHandle`
   rejected with `InvalidStateError: Failed to execute 'requestPersistentHandle'
   on 'XRAnchor': Maximum number of anchors reached!` while
   `session.persistentAnchors` was `["", "", "", "", "", "", "", "", ""]`.
   `deletePersistentAnchor("")` -> `OperationError: Anchor deletion failed.`
   Clearing the origin's site data (CDP `Storage.clearDataForOrigin`) emptied
   it: the next session listed `[""]` and the next handle was granted.

2. Deleting does not appear to free a slot. Two protocols on a freshly cleared
   store, one alignment per session, page reloaded between sessions:
   a) at session start delete every named id the list returns, awaited one at
      a time -> handle refused at the 4th session (listing: 8 blank + 1 named);
   b) at session start delete only the previous session's stored uuid, awaited,
      no other deletes -> named/blank listing grew 1/2, 2/3, 3/4, 4/5 and the
      5th `requestPersistentHandle` was refused with the same message.
   Each accepted delete left one more blank entry, and the named id it deleted
   stayed in the listing. Within a session the list never shrinks after a
   delete (as reported above), and a second delete of the same id in a later
   session rejects with `InvalidStateError: Anchor is already being restored
   or deleted`; once that left the browser unresponsive until a restart.

3. `restorePersistentAnchor(uuid)` returns a new XRAnchor object each call in
   the same session (four consecutive calls, four swaps), which is what lets
   us recover a stuck anchor after a tracking break. Calling `delete()` on a
   restored object appears to erase the persisted record too: the next session
   `restorePersistentAnchor` of that uuid rejected with `OperationError:
   Anchor creation failed.`

Questions: 
(a) in current builds, does `deletePersistentAnchor` free a slot against the cap of 8, or does it leave the blank entry that still counts?
(b) is the blank-entry regression you mentioned the same defect, and does its fix change (a)? 
(c) is `XRAnchor.delete()` on a restored anchor expected to erase the persisted record?
